pkg/operator/watchdog.go

```diff
@@ -0,0 +1,172 @@
+package operator
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/ratelimit"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/dapr/dapr/utils"
+)
+
+const (
+	sidecarContainerName     = "daprd"
+	daprEnabledAnnotationKey = "dapr.io/enabled"
+)
+
+// DaprWatchdog is a controller that periodically polls all pods and ensures that they are in the correct state.
+// This controller only runs on the cluster's leader.
+// Currently, this ensures that the sidecar is injected in each pod, otherwise it kills the pod so it can be restarted.
+type DaprWatchdog struct {
+	enabled           bool
+	interval          time.Duration
+	maxRestartsPerMin int
+
+	client         client.Client
+	restartLimiter ratelimit.Limiter
+}
+
+// NeedLeaderElection makes it so the controller runs on the leader node only.
+// Implements sigs.k8s.io/controller-runtime/pkg/manager.LeaderElectionRunnable .
+func (dw *DaprWatchdog) NeedLeaderElection() bool {
+	return true
+}
+
+// Start the controller. This method blocks until the context is canceled.
+// Implements sigs.k8s.io/controller-runtime/pkg/manager.Runnable .
+func (dw *DaprWatchdog) Start(parentCtx context.Context) error {
+	if !dw.enabled {
+		log.Infof("DaprWatchdog is not enabled")
+		return nil
+	}
+
+	log.Debugf("DaprWatchdog worker started")
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	defer cancel()
+
+	if dw.maxRestartsPerMin > 0 {
+		dw.restartLimiter = ratelimit.New(
+			dw.maxRestartsPerMin,
+			ratelimit.Per(time.Minute),
+			ratelimit.WithoutSlack,
+		)
+	} else {
+		dw.restartLimiter = ratelimit.NewUnlimited()
+	}
+
+	// Use a buffered channel to make sure that there's at most one iteration at a given time
+	// and if an iteration isn't done by the time the next one is scheduled to start, no more iterations are added to the queue
+	workCh := make(chan struct{}, 1)
+	defer close(workCh)
+	firstCompleteCh := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-workCh:
+				dw.listPods(ctx)
+				if firstCompleteCh != nil {
+					close(firstCompleteCh)
+					firstCompleteCh = nil
+				}
+			}
+		}
+	}()
+
+	// Start an iteration right away, at startup, then wait for completion
+	workCh <- struct{}{}
+	<-firstCompleteCh
+
+	// If we only run once, exit when it's done
+	if dw.interval < time.Second {
+		return nil
+	}
+
+	// Repeat on interval
+	t := time.NewTicker(dw.interval)
+	defer t.Stop()
+
+forloop:
+	for {
+		select {
+		case <-ctx.Done():
+			break forloop
+		case <-t.C:
+			log.Debugf("DaprWatchdog worker tick")
+			select {
+			case workCh <- struct{}{}:
+				// Successfully queued another iteration
+				log.Debugf("Added listPods iteration to the queue")
+			default:
+				// Do nothing - there's already an iteration in the queue
+				log.Debugf("There's already an iteration in the listPods queue-not adding another one")
+			}
+		}
+	}
+
+	log.Debugf("DaprWatchdog worker stopping")
```

Comment:
Info.

Also, log when it stops.

Reply:
Done. Stopped needs to be logged here: https://github.com/dapr/dapr/pull/4764/files#diff-76c380a58f117638bd5ab67cbcf2704d7e623b3d3651204a8fcccda177317ebaR66